### PR TITLE
Implement a well-behaved stacktrace printer.

### PR DESCRIPTION
**Pull request by bbockelm**
This implements a stack trace printer which is safe to call
from a signal handler (ROOT's stack trace printer is not async-signal
safe; see 'man 7 signal' for more information about async-signal
safety).

Further, on Linux, this utilizes the clone library call directly;
clone has the advantage of avoiding the pthread_atfork handlers.
We have observed that jemalloc currently registers AS-unsafe atfork
handlers, which would cause a deadlock if we use fork().

The commit addresses real deadlocks observed in the Tier-0.

### CODE REVIEW

**Comment by bbockelm** [comment]
@davidlange6 - I think this should fix the deadlocks observed at the Tier-0.

Basically, the patch is just very careful about avoiding memory allocations and most standard library functions from the context of the signal handler.  Note also I use the CLONE_VM to avoid memory allocation issues in the child process before execv is invoked.

FWIW - I did check the glibc implementation of clone() to validate it is AS-safe.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
Does this work on a Mac?

**Comment by sextonkennedy** [comment]
I asked google the MAC question and it came up with this:
http://stackoverflow.com/questions/24523118/pstack-a-process-on-osx-10-9
which sounds promising...

**Comment by bbockelm** [comment]
@Dr15Jones - It seems the Jenkins failures are unrelated.

@sextonkennedy - I switched from using 'pstack' to using the 'gdb' shipped by CMSSW; this should be available on all our platforms.  The upside of our `gdb` is we ship symbols, so tracebacks have an immense amount of debugging information in them.  The downside of our `gdb` is it appears to take >1GB RAM and 10 minutes to generate a traceback.

**Comment by Dr15Jones** [comment]
please test

**Comment by bbockelm** [comment]
@sextonkennedy - @smuzaffar clarified my comment above via HN.  The very slow GDB will only occur on releases where debug symbols are shipped.  I was doing my testing based on an IB; "normal" resources shouldn't have this issue.

File: FWCore/Services/src/InitRootHandlers.cc [hunk 1]
```diff
@@ -337,5 +414,17 @@ namespace edm {
       s_ignoreWarnings = true;
     }
 
+    void
+    InitRootHandlers::cachePidInfo()
+    {
+      assert(snprintf(pidString_, pidStringLength_-1, "gdb -quiet -p %d 2>&1 <<EOF |\n"
```
Comment (Dr15Jones) [inline FWCore/Services/src/InitRootHandlers.cc]: Don't do real work inside an assert since a compile switch can make the assert go away.

**Comment by Dr15Jones** [comment]
+1

**Comment by bbockelm** [comment]
@davidlange6 @Dr15Jones - what (if any) branches would you like this backported to?

**Comment by bbockelm** [comment]
@davidlange6 - bump on the question above.  Would you like a backport to any other branch?

**Comment by Dr15Jones** [comment]
please test

**Pull request merged**